transportable-udfs-annotation-processor/src/main/java/com/linkedin/transport/processor/TransportProcessor.java

```diff
@@ -106,18 +122,40 @@ private void processElements(Set<? extends Element> elements) {
   private void processUDFClass(TypeElement udfClassElement) {
     debug(String.format("Processing UDF Class: %s", udfClassElement.getQualifiedName()));
 
-    TypeElement closestTopLevelStdUDFInterface = getClosestTopLevelStdUDFInterface(udfClassElement);
-    if (closestTopLevelStdUDFInterface != null) {
-      _udfProperties.addUDF(closestTopLevelStdUDFInterface.getQualifiedName().toString(),
+    if (isValidUDFClass(udfClassElement)) {
+      // At this point we have already verified that TopLevelStdUDF interface is coming from one and only one of the
+      // implemented interfaces
+      TypeElement elementImplementingTopLevelStdUDFInterface =
+          getElementImplementingTopLevelStdUDFInterface(udfClassElement);
+
+      _udfProperties.addUDF(elementImplementingTopLevelStdUDFInterface.getQualifiedName().toString(),
           udfClassElement.getQualifiedName().toString());
     }
   }
 
   /**
-   * Returns a {@link TypeElement} for the closest (in the type hierarchy) class/interface which implements the
-   * {@link TopLevelStdUDF} interface for a given {@link StdUDF} class
+   * Returns the element(class/interface) implementing {@link TopLevelStdUDF} for the given class
+   *
+   * If the class directly implements {@link TopLevelStdUDF}, the class is returned.
+   * If the class indirectly implements {@link TopLevelStdUDF} through an interface, the interface is returned.
+   */
+  private TypeElement getElementImplementingTopLevelStdUDFInterface(TypeElement udfClassElement) {
+    TypeMirror ifaceImplementingTopLevelStdUDFInterface = udfClassElement.getInterfaces()
+        .stream()
+        .filter(iface -> _types.isAssignable(iface, _topLevelStdUDFInterfaceType))
+        .findFirst().get();
+
+    if (_types.isSameType(ifaceImplementingTopLevelStdUDFInterface, _topLevelStdUDFInterfaceType)) {
+      return udfClassElement;
+    } else {
+      return (TypeElement) _types.asElement(ifaceImplementingTopLevelStdUDFInterface);
+    }
+  }
+
+  /**
+   * Performs validations on the type hierarchy to verify UDF overloading rules
    */
-  private TypeElement getClosestTopLevelStdUDFInterface(TypeElement udfClassElement) {
+  private boolean isValidUDFClass(TypeElement udfClassElement) {
```

Comment:
As discussed, this can be further simplified. A UDF can have only one chain from the UDF to TopLevelUDF class, and the last link in the chain should have the implementation of getFunctionName and getFunction Description.

Reply:
As discussed yesterday, we will be checking if there is only one overriding of TopLevelStdUDF methods regardless of leaf/non-leaf, abstract/non-abstract.